azure.go

```diff
@@ -120,87 +140,87 @@ func (p *azureProvider) Verifier() tokenVerifier {
 		return nil, err
 	}
 
-	client := compute.NewVirtualMachineScaleSetsClientWithBaseURI(p.settings.Environment.ResourceManagerEndpoint, subscriptionID)
-	client.Authorizer = authorizer
-	client.Sender = p.httpClient
-	client.AddToUserAgent(userAgent(p.settings.PluginEnv))
-	return client, nil
-}
-
-func (p *azureProvider) MSIClient(subscriptionID string) (msiClient, error) {
-	authorizer, err := p.getAuthorizer()
+	clientOptions := &arm.ClientOptions{
+		ClientOptions: policy.ClientOptions{
+			Transport: transporter{
+				pluginEnv: p.settings.PluginEnv,
+				sender:    p.httpClient,
+			},
+		},
+	}
+	client, err := armcompute.NewVirtualMachineScaleSetsClient(subscriptionID, cred, clientOptions)
 	if err != nil {
 		return nil, err
 	}
 
-	client := msi.NewUserAssignedIdentitiesClientWithBaseURI(p.settings.Environment.ResourceManagerEndpoint, subscriptionID)
-	client.Authorizer = authorizer
-	client.Sender = p.httpClient
-	client.AddToUserAgent(userAgent(p.settings.PluginEnv))
 	return client, nil
 }
 
-func (p *azureProvider) getAuthorizer() (autorest.Authorizer, error) {
-	p.lock.RLock()
-	unlockFunc := p.lock.RUnlock
-	defer func() { unlockFunc() }()
-
-	if p.authorizer != nil && time.Now().Before(p.authorizerExpiration) {
-		return p.authorizer, nil
+func (p *azureProvider) MSIClient(subscriptionID string) (msiClient, error) {
+	cred, err := p.getTokenCredential()
+	if err != nil {
+		return nil, err
 	}
 
-	// Upgrade lock
-	p.lock.RUnlock()
-	p.lock.Lock()
-	unlockFunc = p.lock.Unlock
-
-	if p.authorizer != nil && time.Now().Before(p.authorizerExpiration) {
-		return p.authorizer, nil
+	clientOptions := &arm.ClientOptions{
+		ClientOptions: policy.ClientOptions{
+			Transport: transporter{
+				pluginEnv: p.settings.PluginEnv,
+				sender:    p.httpClient,
+			},
+		},
+	}
+	client, err := armmsi.NewUserAssignedIdentitiesClient(subscriptionID, cred, clientOptions)
+	if err != nil {
+		return nil, err
 	}
 
-	// Create an OAuth2 client for retrieving VM data
-	var authorizer autorest.Authorizer
-	var err error
-	switch {
-	// Use environment/config first
-	case p.settings.ClientSecret != "":
-		config := auth.NewClientCredentialsConfig(p.settings.ClientID, p.settings.ClientSecret, p.settings.TenantID)
-		config.AADEndpoint = p.settings.Environment.ActiveDirectoryEndpoint
-		config.Resource = p.settings.Environment.ResourceManagerEndpoint
-		authorizer, err = config.Authorizer()
+	return client, nil
+}
+
+func (p *azureProvider) getTokenCredential() (azcore.TokenCredential, error) {
+	if p.settings.ClientSecret != "" {
+		cred, err := az.NewClientSecretCredential(p.settings.TenantID, p.settings.ClientID, p.settings.ClientSecret, nil)
 		if err != nil {
 			return nil, err
 		}
-	// By default use MSI
-	default:
-		config := auth.NewMSIConfig()
-		config.Resource = p.settings.Environment.ResourceManagerEndpoint
-		authorizer, err = config.Authorizer()
+
+		return cred, nil
+	} else {
+		cred, err := az.NewManagedIdentityCredential(nil)
```

Comment:
I _think_ we may want to set the TenantID here? I think that is only really useful for local development where you don't want to pass a clientID and clientSecret and you are logged in via cli `az login`. Or maybe we want to always require clientID and clientSecret for local development?

Reply:
I'm still a little unsure about what the correct behavior here is, because upon a closer look I do see the internal autorest auth package using a ClientID while trying to create credentials: https://github.com/Azure/go-autorest/blob/main/autorest/azure/auth/auth.go#L745, so if that behavior is to be emulated then I should pass a ClientID in here.

I was led to believe that we prefer to use managed identity credentials (or MSI configs in autorest terms) only when the ClientID, ClientSecret and TenantID are all empty. This was partly because of how the code is implemented in Azure Auth main [depicted here](https://github.com/hashicorp/vault-plugin-auth-azure/blob/main/azure.go#L184-L200), and also in Azure Secrets main [shown here](https://github.com/hashicorp/vault-plugin-secrets-azure/blob/main/provider.go#L97-L104). From the code it looks MSI configs are created only if the environment variables are empty

Reply:
Yeah, I think it makes sense to keep it as you have it here. That should keep the behavior consistent with how it has always been.